libs/pika/async_mpi/tests/unit/mpi_ring_async_sender_receiver.cpp

```diff
@@ -0,0 +1,252 @@
+
+        // Ring send/recv around N ranks
+        // Rank 0      : Send then Recv
+        // Rank 1->N-1 : Recv then Send
+
+        pika::mpi::experimental::executor exec(MPI_COMM_WORLD);
+
+        std::vector<int> tokens(iterations, -1);
+
+        pika::chrono::high_resolution_timer t;
+
+        const std::uint64_t inner_iterations = 100;
+        const std::uint64_t outer_iterations = iterations / inner_iterations;
+        std::atomic<std::uint64_t> counter;
+
+        std::uint64_t tag;
+        for (std::uint64_t j = 0; (j != outer_iterations); ++j)
+        {
+            counter = inner_iterations;
+            for (std::uint64_t i = 0; (i != inner_iterations); ++i)
+            {
+                tag = i + j * inner_iterations;
+                tokens[tag] = (rank == 0) ? 1 : -1;
+                int rank_from = (size + rank - 1) % size;
+                int rank_to = (rank + 1) % size;
+
+                // all ranks pre-post a receive
+                auto s1 =
+                    mpi::transform_mpi(ex::just(&tokens[tag], 1, MPI_INT,
+                                           rank_from, tag, MPI_COMM_WORLD),
+                        MPI_Irecv);
+
+                // when that receive completes, send the message on to the next rank
+                auto s2 = s1 |
+                    pika::execution::experimental::then(
+                        [=, &tokens, &counter](int /*result*/) {
+                            msg_recv(rank, size, rank_to, rank_from,
+                                tokens[tag], tag);
+                            if (rank > 0)
+                            {
+                                // send the incremented token to the next rank
+                                ++tokens[tag];
+                                auto s3 = mpi::transform_mpi(
+                                    ex::just(&tokens[tag], 1, MPI_INT, rank_to,
+                                        tag, MPI_COMM_WORLD),
+                                    MPI_Isend);
+
+                                // when the send completes
+                                auto s4 = s3 |
+                                    pika::execution::experimental::then(
+                                        [=, &tokens, &counter](int /*result*/) {
+                                            msg_send(rank, size, rank_to,
+                                                rank_from, tokens[tag], tag);
+                                            // ranks > 0 are done after sending token
+                                            --counter;
+                                        });
+                                ex::start_detached(s4);
+                            }
+                            else
+                            {
+                                // rank 0 is done when it receives its token
+                                --counter;
+                            }
+                        });
+                ex::start_detached(s2);
```

Comment:
Please also move the senders to the next adaptor here to avoid unnecessary copies.

I'm also wondering if it wouldn't be clearer to write this without nesting the send inside a `then` after the recv? At least that would avoid allocating two shared states with `start_detached` which is unnecessary.

It would also be nice to check the status from the MPI senders, perhaps with a new helper `check_mpi_error` like we have for CUDA?

Reply:
Done. probably